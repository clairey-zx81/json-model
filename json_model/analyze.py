#
# Analyses
#
import re
from .mtypes import UnknownModel, ModelPath, ModelType, ModelFilter
from .utils import log, CONST_RE, is_regex
from .recurse import recModel, allFlt, noRwt
from .model import JsonModel

# JsonModel = typing.NewType("JsonModel", None)
type Constants = bool|int|float|str
type ModelRef = list[str]

def references(jm: JsonModel) -> dict[str, set[int]]:
    """Get direct references between models.

    local name in model -> set of model ids that are referenced
    """
    assert jm._is_root

    def directRefs(model: ModelType) -> set[int]:
        """Recursive collect references."""

        refs: set[int] = set()

        def drRwt(m: ModelType, p: ModelPath) -> ModelType:
            if isinstance(m, str) and jm._isRef(m):
                target = jm.resolveRef(m, p)
                refs.add(target._id)
            return m

        recModel(model, allFlt, drRwt, True)

        return refs

    contains = {}

    for name, model in jm._defs.items():
        contains[name] = directRefs(model._model)

    contains[""] = directRefs(jm._model)

    return contains

# def reachable(self):
#     """Build reachable defs."""
#
#     # should consider **all** loaded models!?
#     contains = self.references()
#     reached = copy.deepcopy(contains)
#
#     # simplistic transitive closure computation
#     changed = True
#     while changed:
#         changed = False
#         for ref, refs in reached.items():
#             for r in refs:
#                 if r in reached and not reached[r].issubset(refs):
#                     changed = True
#                     refs.update(reached[r])
#
#     return reached

def check(jm: JsonModel, assertion: ModelFilter, what: str = "?", short: bool = False) -> bool:
    """Check an arbitrary local assertion on a model."""

    okay = True

    def checkFlt(model: ModelType, path: ModelPath) -> bool:
        nonlocal okay
        if not assertion(model, path):
            log.debug(f"{jm._id}: check {what} failed at {path}")
            okay = False
        return okay if short else True

    recModel(jm._model, checkFlt, noRwt, True)

    return okay

# FIXME must tell why it is unhappy!
def valid(jm: JsonModel, path: ModelPath = [], root: bool = True) -> bool:
    """Check JSON Model structural (and slightly more) validity."""

    is_valid = True

    def finiteRef(model: str) -> bool:
        ref, recref, recid, ljm = model, [], [], jm
        while isinstance(ref, str) and jm._isRef(ref) and ref not in recref:
            ljm = ljm.resolveRef(ref, path + recref)
            recref.append(ref)
            recid.append(ljm._id)
            ref = ljm._model
        finite = not jm._isRef(ref) or ref == "$#"  # Hmmm…
        if not finite:
            log.debug(f"infinite recursion on {model}: {recref} ({recid})")
        return finite

    def validFlt(model: ModelType, path: ModelPath) -> bool:
        nonlocal is_valid
        match model:
            case None:
                pass
            case bool():
                is_valid &= model
            case int():
                is_valid &= model in (-1, 0, 1)
            case float():
                is_valid &= model in (-1.0, 0.0, 1.0)
            case str():
                if model == "":
                    pass
                elif model[0] == "=":
                    is_valid &= re.match(CONST_RE, model) is not None
                elif model[0] == "/":
                    try:
                        pattern, ropts = model[1:].rsplit("/", 1)
                        is_valid &= (ropts == "" or ropts.isalpha()) and is_regex(pattern)
                    except Exception:
                        is_valid = False
                elif model[0] == "$":
                    pass
                else:  # TODO more checks
                    pass
            case list():
                pass
            case dict():
                if "#" in model:
                    is_valid &= isinstance(model["#"], str)
                if "@" in model:
                    if "!" in model:
                        is_valid &= isinstance(model["!"], bool)
                    for op in ("=", "!=", "<", "<=", ">", ">="):
                        if op in model:
                            is_valid &= isinstance(model[op], (int, float, str))
                    # TODO reject any other props! TODO allow .mo and other extensions?
                    for k in model.keys():
                        is_valid &= \
                            k in ("@", "!", "=", "!=", "<", "<=", ">=", ">", "#", "$", "%", "~")
                elif "|" in model:
                    is_valid &= isinstance(model["|"], list)
                    # no other keys in recurse
                elif "^" in model:
                    is_valid &= isinstance(model["^"], list)
                    # no other keys in recurse
                elif "&" in model:
                    is_valid &= isinstance(model["&"], list)
                    # no other keys in recurse
                elif "+" in model:
                    is_valid &= isinstance(model["+"], list)
                    # no other keys in recurse
                else:  # check object
                    props = set()  # property name collisions
                    for p, m in model.items():
                        if p == "#":
                            continue
                        is_valid &= isinstance(p, str)
                        if p and p[0] not in ("$", "/"):
                            name = p[1:] if p[0] in ("?", "_", "!") else p
                            is_valid &= name not in props
                            props.add(name)
                        # more checks on p if p[0] == "$"
            case _:
                is_valid = False
        return True

    if is_valid and jm._defs:
        for name, jmr in jm._defs.items():
            is_valid &= finiteRef("$" + name)

    if is_valid:
        try:
            check(jm, validFlt, "JSON Model Structural Validity")
        except AssertionError as e:
            log.error(e, exc_info=jm._debug > 0)
            is_valid = False
            if jm._debug:
                raise

    # TODO also check other instances?!

    return is_valid



# ultimate type of predefs
_UTYPE = {
    "$ANY": None, "$NONE": None,
    "$NULL": type(None),
    "$BOOL": bool, "$BOOLEAN": bool,
    "$I32": int, "$U32": int, "$I64": int, "$U64": int, "$INT": int, "$INTEGER": int,
    "$F32": float, "$F64": float, "$FLOAT": float, "$NUMBER": float,
    "$STRING": str,
    "$URL": str, "$URI": str, "$EMAIL": str,
    "$REGEX": str, "$EXREG": str,
    "$DATE": str, "$TIME": str, "$DATETIME": str,
    "$UUID": str, "$JSON": str,
}


# TODO return a set of types?
def ultimate_type(jm: JsonModel, model: ModelType) -> type|None:
    """Get the utimate type by following definitions.

    Return _None_ if unknown, eg several are possible.
    Note: this is also return of the type system is not feasible, eg "&(int, str)"

    FIXME recursion
    """
    match model:
        case str():
            if model == "" or model[0] not in ("$", "="):
                return type(model)
            elif model[0] == "=":  # constants
                return (type(None) if model == "=null" else
                        bool if model in ("=true", "=false") else
                        float if "." in model else
                        int)
            elif model in _UTYPE:
                return _UTYPE[model]
            else:
                m = jm.resolveRef(model, ["<?>"])  # type: ignore
                return ultimate_type(m, m._model)
        case dict():
            if "@" in model:
                return ultimate_type(jm, model["@"])
            elif "|" in model:
                assert isinstance(models := model["|"], list)
                types = set(ultimate_type(jm, m) for m in models)
                if len(types) == 1:
                    return types.pop()
                else:
                    return None
            elif "^" in model:
                assert isinstance(models := model["^"], list)
                types = set(ultimate_type(jm, m) for m in models)
                if len(types) == 1:
                    return types.pop()
                else:
                    return None
            elif "&" in model:
                assert isinstance(models := model["&"], list)
                types = set(ultimate_type(jm, m) for m in models)
                if len(types) == 1:
                    return types.pop()
                else:
                    # possibly not feasible
                    return None
            else:
                return type(model)  # dict
        case _:  # None, bool, int, float, list
            return type(model)


# model predef to standard model
_UMODEL = {
    "$BOOL": True, "$NULL": None,
    "$I32": -1, "$U32": 0, "$I64": -1, "$U64": 0, "$INTEGER": -1,
    "$F32": -1.0, "$F64": -1.0, "$NUMBER": -1.0,
    "$STRING": "",
    # stop recursion on (some) predefs
    "$ANY": "$ANY",
    "$NONE": "$NONE",
    "$URL": "$URL",
    "$REGEX": "$REGEX",
    "$EXREG": "$EXREG",
    "$DATE": "$DATE",
    "$TIME": "$TIME",
    "$DATETIME": "$DATETIME",
    "$UUID": "$UUID",
    "$URI": "$URI",
    "$EMAIL": "$EMAIL",
    "$JSON": "$JSON",
    # to be continued…
}

def ultimate_model(jm: JsonModel, model: ModelType, constrained=True, strict=False) -> ModelType:
    """Look for the real model, beyond references."""
    log.debug(f"ultimate_model on {model}")
    match model:
        case str():
            if model == "" or model[0] not in ("$", "="):  # / returns as is
                return model
            elif model[0] == "=":
                return model
            else:  # follow definition if possible
                assert model[0] == "$"
                if model in _UMODEL:
                    m = _UMODEL[model]
                else:
                    m = jm.resolveRef(model, [])._model
                # handle some predefs (again)
                if isinstance(m, str) and m in _UMODEL:
                    return _UMODEL[m]
                # else try recursing
                return ultimate_model(jm, m) if m != UnknownModel else m  # type: ignore
        case dict():
            if "@" in model:
                if strict:
                    if set(model.keys()).issubset(["@", "#", "~", "%", "$"]):
                        return ultimate_model(jm, model["@"])
                    else:
                        return UnknownModel  # type: ignore
                elif constrained:
                    return ultimate_model(jm, model["@"])
                else:
                    return UnknownModel  # type: ignore
            elif "|" in model:
                return UnknownModel  # type: ignore
            else:
                return model
        case _:
            return model

def evalModel(jm: JsonModel, model: ModelType):
    """Tell if an ultimate model value has a constant."""
    # FIXME should it detect @ eq?
    v = ultimate_model(jm, model)
    if v == UnknownModel:
        # FIXME ?!
        return None
    tv = type(v)
    if tv is str:
        assert isinstance(v, str)  # useless pyright hint
        if v == "":
            return None  # FIXME
        elif v == "=null":
            return None
        elif re.search(r"^=(true|false)$", v) is not None:
            return True if v == "=true" else False
        elif re.search(r"^=-?\d+$", v) is not None:
            return int(v[1:])
        elif re.search(r"^=", v) is not None:
            return float(v[1:])
        elif v[0] == "_":
            return v[1:]
        else:
            return v
    else:
        return None

def disjunct_analyse(jm: JsonModel, model: ModelType, mpath: ModelPath) -> \
        tuple[str, type, list, list]|None:
    """Return the optimized check function if possible."""
    # FIXME if there is a ^, the preprocessor will have detected the discriminant
    # and turned it into a |.
    assert isinstance(model, dict) and "|" in model
    # first filter out
    utype = ultimate_type(jm, model)
    if utype is not dict:
        log.debug(f"ultimate type not a dict: {utype}")
        return None
    # get models
    assert isinstance(models := model["|"], list)
    models = [ultimate_model(jm, m) for m in models]
    # should not happen, just in case
    if len(models) < 2:
        log.debug("not enough models for a disjunction")
        return None
    # we may have some doubts yet (eg | in |)
    if any(filter(lambda m: not isinstance(m, dict), models)):
        log.debug("some models are not objects")
        return None
    # only objects: collect their direct mandatory properties
    all_props: list[set[str]] = [
        set(k for k in m.keys() if k and k[0] not in ("$", "?", "/"))
        for m in models if isinstance(m, dict)
    ]
    # get cleaned props and their constant values
    all_const_props: list[dict[str, Constants]] = []
    for props, model in zip(all_props, models):
        consts: dict[str, Constants] = {}
        assert isinstance(model, dict)
        for prop in props:
            val = evalModel(jm, model[prop])
            if val is not None:
                key = prop[1:] if prop[0] in ("_", "!") else prop
                consts[key] = val
        all_const_props.append(consts)
    # tag candidates must:
    # - appear in all alternate models
    candidates: set[str] = set(all_const_props[0].keys())
    for props in all_const_props[1:]:
        candidates.intersection_update(props.keys())
    if not candidates:
        log.debug("no property with constant values")
        return None
    # - have the same type
    candidates_typed: set[str] = set()
    for prop in candidates:
        if len(set(type(consts[prop]) for consts in all_const_props)) == 1:
            candidates_typed.add(prop)
    if not candidates_typed:
        log.debug("no proprety with same type constants")
        return None
    # - have distinct constant values
    candidates_distinct: set[str] = set()
    for prop in candidates_typed:
        if len(set(consts[prop] for consts in all_const_props)) == len(all_const_props):
            candidates_distinct.add(prop)
    if not candidates_distinct:
        log.debug("no property with distinct values")
        return None
    if len(candidates_distinct) > 1:
        log.warning(f"several disjunctive properties: {candidates_distinct}")
    # one tag with one type and only constants found!
    tag_name = candidates_distinct.pop()
    tag_type = ultimate_type(jm, all_const_props[0][tag_name])
    assert tag_type in (bool, int, float, str)

    return tag_name, tag_type, models, all_const_props
