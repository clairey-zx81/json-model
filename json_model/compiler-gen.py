# generate python source code for a model

import sys
import re
import json
import logging

from utils import ModelType, ValueType, ModelError, openfiles

logging.basicConfig()
log = logging.getLogger("gen")
log.setLevel(logging.DEBUG)

Line = tuple[int, str]
Code = list[Line]

# TODO encode name if needed
_is_simple_name = re.compile(r"^[_a-zA-Z][_a-zA-Z0-9]*$").search

_PREDEFS = {
    "ANY": lambda _v: "True",
    "NONE": lambda _v: "False",
    "STRING": lambda v: f"isinstance({v}, str)",
}

class SourceCode():

    def __init__(self, model: ModelType):
        self._nvars: dict[str, int] = {}
        self._names: set[str] = set()
        self._model: ModelType = model
        self._code: Code = []
        self._compileRoot(model)

    def __str__(self):
        return "\n".join(("    " * line[0] + line[1]) for line in self._code)

    def var(self, prefix: str):
        if prefix not in self._nvars:
            self._nvars[prefix] = 0
        v = f"{prefix}{self._nvars[prefix]}"
        self._nvars[prefix] += 1
        return v

    def esc(self, string: str):
        return string.translate({"\"": "\\\"", "\\": "\\\\"})

    def line(self, indent: int, line: str):
        self._code.append((indent, line))

    def _compileModel(self, model: ModelType, res: str, val: str, indent: int, mpath: str):
        log.debug(f"model={model} res={res} val={val} indent={indent} mpath={mpath}")
        if model is None:
            self.line(indent, f"{res} = {val} is None")
        elif isinstance(model, bool):
            self.line(indent, f"{res} = isinstance({val}, bool)")
        elif isinstance(model, int):
            expr = f"isinstance({val}, int) and not isinstance({val}, bool)"
            if model == -1:
                pass
            elif model == 0:
                expr += f" and {val} >= 0"
            elif model == 1:
                expr += f" and {val} >= 1"
            else:
                raise ModelError(f"unexpected int value {model} at {mpath}")
            self.line(indent, res + " = " + expr)
        elif isinstance(model, int):
            expr = "isinstance({val}, int) and not isinstance({val}, bool)"
            if model == -1:
                pass
            elif model == 0:
                expr += f" and {val} >= 0"
            elif model == 1:
                expr += f" and {val} >= 1"
            else:
                raise ModelError(f"unexpected int value {model} at {mpath}")
            self.line(indent, res + " = " + expr)
        elif isinstance(model, float):
            expr = f"isinstance({val}, float)"
            if model == -1.0:
                pass
            elif model == 0.0:
                expr += f" and {val} >= 0.0"
            elif model == 1.0:
                expr += f" and {val} >= 1.0"
            else:
                raise ModelError(f"unexpected float value {model} at {mpath}")
            self.line(indent, res + " = " + expr)
        elif isinstance(model, str):
            expr = f"isinstance({val}, str)"
            if model == "":
                self.line(indent, f"{res} = " + expr)
            elif model[0] == "_":
                self.line(indent, f"{res} = " + expr + f" and {val} == {esc(model[1:])}")
            elif model[0] == "$":
                name = model[1:]
                if name in _PREDEFS:
                    self.line(indent, f"{res} = " + _PREDEFS[name](val))
                else:
                    self.line(indent, f"{res} = check_{name}({val}, path)")
            elif model[0] == "/":
                raise NotImplementedError("/ handling")
            else:  # simple string
                self.line(indent, f"{res} = " + expr + f" and {val} == \"{esc(model)}\"")
        elif isinstance(model, list):
            expr = f"isinstance({val}, list)"
            if len(model) == 0:
                expr += f" and len({val}) == 0"
            elif len(model) == 1:
                item = self.var("i")
                self.line(indent, res + " = " + expr)
                self.line(indent, f"if {res}:")
                # TODO enumerate
                self.line(indent+1, f"for {item} in {val}:")
                self._compileModel(model[0], res, item, indent+2, mpath + "[0]")
                self.line(indent+2, f"if not {res}: break")
            else:
                raise NotImplementedError("tuple check")
        elif isinstance(model, dict):
            if "@" in model:
                self._compile(m["@"], res, val, indent, mpath + ".@")
                # TODO constraints
                raise NotImplementedError("@ check")
            if "|" in model:
                # TODO discriminant optimization
                lpath = mpath + ".|"
                models = model["|"]
                if not models:
                    self.line(indent, f"{res} = False")
                for i, m in enumerate(models):
                    if i:
                        self.line(indent + i - 1, f"if not {res}:")
                    self._compileModel(m, res, val, indent + i, lpath + f"[{i}]")
            elif "&" in model:
                lpath = mpath + ".&"
                models = model["&"]
                if not models:
                    self.line(indent, f"{res} = True")
                for i, m in enumerate(models):
                    if i:
                        self.line(indent + i - 1, f"if {res}:")
                    self._compileModel(m, res, val, indent + i, lpath + f"[{i}]")
            elif "^" in model:
                raise NotImplementedError("^ check")
            else:
                # TODO separate properties
                raise NotImplementedError("obj check")
        else:
            raise ModelError(f"unexpected model type: {type(model).__name__}")

    def _compileName(self, name: str, model: ModelType, mpath: str):
        assert _is_simple_name(name)
        self.line(0, f"def check_{name}(value: ValueType, path: str) -> bool:")
        self._compileModel(model, "result", "value", 1, mpath)
        self.line(1, "return result")
        self.line(0, "")

    def _compileRoot(self, model: ModelType):
        if "%" in model:
            for name, mod in model.items():
                self._compileName(name, f"$.%.{name}")
        self._compileName("main", model, "$")
        
def compile(model: ModelType) -> SourceCode:
    return SourceCode(model)

for fn, fh in openfiles(sys.argv[1:]):
    try:
        model = json.load(fh)
        print(compile(model))
    except Exception as e:
        log.error(f"{fn}: {e}")        
