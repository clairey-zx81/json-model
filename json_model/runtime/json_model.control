comment = 'JSON Model PL/pgSQL Runtime - JSON value validation'
default_version = '1.0'
