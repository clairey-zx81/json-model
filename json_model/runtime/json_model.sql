-- complain if script is sourced in psql, rather than via CREATE EXTENSION
-- \echo Use "CREATE EXTENSION json_model;" to load this file. \quit

CREATE OR REPLACE FUNCTION jm_call(fun TEXT, val JSONB, vpath TEXT, rep TEXT)
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  res BOOLEAN;
BEGIN
  EXECUTE FORMAT('SELECT %s($1, $2, $3)', fun)
    INTO res
    USING val, vpath, rep;
  RETURN res;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION jm_is_valid_date(val TEXT, vpath TEXT, rep TEXT)
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
    PERFORM val::DATE;
    RETURN TRUE;
EXCEPTION
    WHEN data_exception THEN
        RETURN FALSE;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION jm_is_valid_datetime(val TEXT, vpath TEXT, rep TEXT)
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
    PERFORM val::TIMESTAMP;
    RETURN TRUE;
EXCEPTION
    WHEN data_exception THEN
        RETURN FALSE;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION jm_is_valid_regex(val TEXT, vpath TEXT, rep TEXT)
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
    PERFORM '' ~ val;
    RETURN TRUE;
EXCEPTION
    WHEN data_exception THEN
        RETURN FALSE;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION jm_is_valid_url(val TEXT, vpath TEXT, rep TEXT)
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
    RETURN val ~ E'^((https?|file)://.*|\\.)';
END;
$$ LANGUAGE plpgsql;
