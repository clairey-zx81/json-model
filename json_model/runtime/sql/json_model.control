comment = 'JSON Model PL/pgSQL Runtime - JSON value validation'
default_version = '2.0'
