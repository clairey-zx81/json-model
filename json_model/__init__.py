#
# convenient types, classes and functions to process models
#
from .mtypes import Jsonable
from .runtime.types import Report, EntryCheckFun
from .model import JsonModel
from .script import (
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
)

__ALL__ = [
    Jsonable, Report, EntryCheckFun, JsonModel,
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
]
