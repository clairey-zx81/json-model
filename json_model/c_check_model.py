#
# check some json files against a model
#
# usage: $0 model.json file.json â€¦

import sys
import json
import logging
from json_model.compiler import compileModel

def c_check_model():

    assert len(sys.argv) >= 2 

    logging.basicConfig(level=logging.INFO)
    log = logging.getLogger("model")

    with open(sys.argv[1]) as f:
        checkModel = compileModel(json.load(f))

    for fn in sys.argv[2:]:
        valid = False
        try:
            with open(fn) as f:
                valid = checkModel(json.load(f))
                print(f"{fn}: {valid}")
                if not valid:
                    log.info(f"failures: {checkModel._reasons}")
        except Exception as e:
            print(f"{fn}: error")
            log.error(f"{fn}: {e}")
