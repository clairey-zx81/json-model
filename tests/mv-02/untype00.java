//
// Generated by JSON Model Compiler version 2
// see https://github.com/clairey-zx81/json-model
//

import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

@SuppressWarnings("unchecked")
public class untype00 extends ModelChecker
{
    static public final String VERSION = "2";

    public Map<String, Checker> untype00_map_pmap;

    // check $ (.)
    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res;
        // constraint is either on value or on length
        // .
        // .'@'
        // .'@'.'|'.0
        res = json.isInteger(val) && json.asLong(val) >= 1;
        if (! res)
        {
            if (rep != null) rep.addEntry("not a 1 strict int [.'@'.'|'.0]", path);
        }
        if (! res)
        {
            // .'@'.'|'.1
            res = json.isString(val);
            if (! res)
            {
                if (rep != null) rep.addEntry("unexpected string [.'@'.'|'.1]", path);
            }
        }
        if (res)
        {
            if (rep != null) rep.clearEntries();
        }
        else
        {
            if (rep != null) rep.addEntry("no model matched [.'@'.'|']", path);
        }
        if (res)
        {
            res = rt.check_constraint(val, json_model.Runtime.Operator.GE, json.safeJSON("10"), path, rep);
            if (! res)
            {
                if (rep != null) rep.addEntry("constraints failed [.]", path);
            }
        }
        return res;
    }
    

    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            untype00_map_pmap = new HashMap<String, Checker>();
            untype00_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            untype00_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return untype00_map_pmap.get(name);
    }

    public Set<String> models()
    {
        return untype00_map_pmap.keySet();
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new untype00();
        Main.main("untype00", checker, VERSION, args);
    }
}
