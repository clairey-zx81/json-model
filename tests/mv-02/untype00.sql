--
-- Generated by JSON Model Compiler version 2.0b0
-- see https://github.com/clairey-zx81/json-model
--
-- JSON_MODEL_VERSION is 2.0b0
CREATE EXTENSION IF NOT EXISTS json_model;

-- check $ (.)
CREATE OR REPLACE FUNCTION json_model_1(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  res bool;
BEGIN
  -- constraint is either on value or on length
  -- .
  -- .'@'
  -- .'@'.'|'.0
  res := JSONB_TYPEOF(val) = 'number' AND (val)::INT8 = (val)::FLOAT8 AND (val)::INT8 >= 1;
  IF NOT res THEN
    -- .'@'.'|'.1
    res := JSONB_TYPEOF(val) = 'string';
  END IF;
  IF res THEN
    res := jm_check_constraint(val, '>=', 10, path, rep);
  END IF;
  RETURN res;
END;
$$ LANGUAGE PLpgSQL;

CREATE OR REPLACE FUNCTION check_model_map(name TEXT)
RETURNS TEXT STRICT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  map JSONB := JSONB '{"":"json_model_1"}';
BEGIN
  RETURN map->>name;
END;
$$ LANGUAGE plpgsql;

--
-- constant maps initialization
--
TRUNCATE jm_constant_maps;


--
-- JSON Model checking entry point
--
-- TODO INOUT rep?
CREATE OR REPLACE FUNCTION check_model(val JSONB, name TEXT, rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  fun TEXT;
BEGIN
  fun := check_model_map(name);
  IF fun IS NULL THEN
    RAISE EXCEPTION 'model for % not found', name;
  END IF;
  RETURN jm_call(fun, val, NULL, rep);
END;
$$ LANGUAGE plpgsql;
