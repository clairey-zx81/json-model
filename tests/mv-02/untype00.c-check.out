untype00.values.json[0]: PASS
untype00.values.json[1]: PASS
untype00.values.json[2]: FAIL (.: constraints failed at [$]) (.: invalid int constraint)
untype00.values.json[3]: FAIL (.: constraints failed at [$]) (.: invalid int constraint)
untype00.values.json[4]: FAIL (.: constraints failed at [$]) (.: no model matched [$.'@'.'|']) (.: unexpected string [$.'@'.'|'.1]) (.: not a 1 strict int [$.'@'.'|'.0])
untype00.values.json[5]: FAIL (.: constraints failed at [$]) (.: no model matched [$.'@'.'|']) (.: unexpected string [$.'@'.'|'.1]) (.: not a 1 strict int [$.'@'.'|'.0])
untype00.values.json[6]: FAIL (.: constraints failed at [$]) (.: no model matched [$.'@'.'|']) (.: unexpected string [$.'@'.'|'.1]) (.: not a 1 strict int [$.'@'.'|'.0])
untype00.values.json[7]: FAIL (.: constraints failed at [$]) (.: no model matched [$.'@'.'|']) (.: unexpected string [$.'@'.'|'.1]) (.: not a 1 strict int [$.'@'.'|'.0])
untype00.values.json[8]: FAIL (.: constraints failed at [$]) (.: no model matched [$.'@'.'|']) (.: unexpected string [$.'@'.'|'.1]) (.: not a 1 strict int [$.'@'.'|'.0])
untype00.values.json[9]: FAIL (.: constraints failed at [$]) (.: no model matched [$.'@'.'|']) (.: unexpected string [$.'@'.'|'.1]) (.: not a 1 strict int [$.'@'.'|'.0])
untype00.values.json[10]: FAIL (.: constraints failed at [$]) (.: no model matched [$.'@'.'|']) (.: unexpected string [$.'@'.'|'.1]) (.: not a 1 strict int [$.'@'.'|'.0])
